contracts/src/registry/NameWrapperRegistry.sol

```diff
@@ -0,0 +1,247 @@
+    }
+    
+    /**
+     * @dev Allow the owner to renew the name.
+     * 
+     * See v1 fuse: CAN_EXTEND_EXPIRY
+     */
+    function allowOwnerToRenew(uint256 tokenId) public onlyRole(_tokenIdContext(tokenId), EMANCIPATOR_ROLE) {
+        bytes32 tokenIdContext = _tokenIdContext(tokenId);
+        _grantRole(tokenIdContext, RENEW_ROLE, ownerOf(tokenId));
+    }
+
+
+    /**
+     * @dev Prevent the name from being renewed.
+     * 
+     * See v1 fuse: CANNOT_SET_TTL
+     */
+    function lockRenewals(uint256 tokenId) public onlyRole(_tokenIdContext(tokenId), EMANCIPATED_OWNER_ROLE) {
+        bytes32 tokenIdContext = _tokenIdContext(tokenId);
+        _revokeRoleAssignments(tokenIdContext, RENEW_ROLE);
+    }
+
+    /**
+     * @dev Prevent the name from being transferred.
+     * 
+     * See v1 fuse: CANNOT_TRANSFER
+     */
+    function lockTransfers(uint256 tokenId) public onlyRole(_tokenIdContext(tokenId), EMANCIPATED_OWNER_ROLE) {
+        bytes32 tokenIdContext = _tokenIdContext(tokenId);
+        transferLock[tokenIdContext] = true;
+    }
+
+    /**
+     * @dev Renounce the emancipated owner role.
+     * 
+     * See v1 fuse: CANNOT_BURN_FUSES
+     */
+    function renounceEmancipatedOwnerRole(uint256 tokenId) public onlyRole(_tokenIdContext(tokenId), EMANCIPATED_OWNER_ROLE) {
+        bytes32 tokenIdContext = _tokenIdContext(tokenId);
+        renounceRole(tokenIdContext, EMANCIPATED_OWNER_ROLE, msg.sender);
+    }
+
+
+    function nameData(uint256 tokenId) external view returns (uint64 expiry, uint32 flags) {
+        (, uint96 _flags) = datastore.getSubregistry(tokenId);
+        return (_extractExpiry(_flags), uint32(_flags));
+    }
+
+    function setFlags(uint256 tokenId, uint96 flags)
+        external
+        onlyTokenOwner(tokenId)
+        returns (uint256 newTokenId)
+    {
+        uint96 newFlags = _setFlags(tokenId, flags);
+        newTokenId = (tokenId & ~uint256(FLAGS_MASK)) | (newFlags & FLAGS_MASK);
+        if (tokenId != newTokenId) {
+            address owner = ownerOf(tokenId);
+            _burn(owner, tokenId, 1);
+            _mint(owner, newTokenId, 1, "");
+
+            // NOTE: the role assignments can stay as they are since the 
+            // token's role assignment context value remains unchanged.
+            // @see _tokenIdContext
+        }
+    }
+
+    function supportsInterface(bytes4 interfaceId) public view override(BaseRegistry, EnhancedAccessControl) returns (bool) {
+        return interfaceId == type(IRegistry).interfaceId || super.supportsInterface(interfaceId);
+    }
+
+    function getSubregistry(string calldata label) external view virtual override returns (IRegistry) {
+        (address subregistry, uint96 flags) = datastore.getSubregistry(uint256(keccak256(bytes(label))));
+        uint64 expires = _extractExpiry(flags);
+        if (expires <= block.timestamp) {
+            return IRegistry(address(0));
+        }
```

Comment:
I see this reused in several places; we should use an accessor function to remove the duplication.